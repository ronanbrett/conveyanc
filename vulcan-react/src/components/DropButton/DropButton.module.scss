@import "src/styles/vars";



.DropButton {

}