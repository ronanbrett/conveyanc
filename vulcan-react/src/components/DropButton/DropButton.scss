@import "src/styles/vars";

.DropButton {
  border-radius: 0 !important;
  border-color: var(--border-color) !important;

  &.focus {
    border-color: var(--highlight-color) !important;
  }
  &.opened {
    border-color: var(--highlight-color) !important;
    border-bottom-right-radius: 0;
    border-bottom-left-radius: 0;
  }
}
