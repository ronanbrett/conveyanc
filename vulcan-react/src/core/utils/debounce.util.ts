export const debounce = (cb: any, timer: number) => {
  let timeout: any;
  return (...args: any) => {
    // @ts-ignore
    const context = this;
    clearTimeout(timeout);
    timeout = setTimeout(() => cb.apply(context, args), timer);
  };
};
