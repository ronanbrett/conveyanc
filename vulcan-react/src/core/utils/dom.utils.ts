export const findScrollParent = (
  element: HTMLElement,
  horizontal?: boolean
) => {
  let result;
  if (element) {
    let parent = element.parentNode as HTMLElement;
    while (!result && parent && parent.getBoundingClientRect) {
      const rect = parent.getBoundingClientRect();
      // 10px is to account for borders and scrollbars in a lazy way
      if (horizontal) {
        if (rect.width && parent.scrollWidth > rect.width + 10) {
          result = parent;
        }
      } else if (rect.height && parent.scrollHeight > rect.height + 10) {
        result = parent;
      }
      parent = parent.parentNode as HTMLElement;
    }
    // last scrollable element will be the document
    // if nothing else is scrollable in the page
    if (!result) {
      result = document;
    } else if (result.tagName.toLowerCase() === "body") {
      result = document;
    }
  }
  return result;
};

export const findScrollParents = (
  element: HTMLElement,
  horizontal?: boolean
) => {
  const result = [];
  if (element) {
    let parent = element.parentNode as HTMLElement;
    while (parent && parent.getBoundingClientRect) {
      const rect = parent.getBoundingClientRect();
      // 10px is to account for borders and scrollbars in a lazy way
      if (horizontal) {
        if (rect.width && parent.scrollWidth > rect.width + 10) {
          result.push(parent);
        }
      } else if (rect.height && parent.scrollHeight > rect.height + 10) {
        result.push(parent);
      }
      parent = parent.parentNode as HTMLElement;
    }
    // last scrollable element will be the document
    // if nothing else is scrollable in the page
    if (result.length === 0) {
      result.push(document);
    } else if (result[0].tagName.toLowerCase() === "body") {
      result.length = 0;
      result.push(document);
    }
  }
  return result;
};

export const containsFocus = (node: Element) => {
  let element = document.activeElement;
  while (element) {
    if (element === node) break;
    element = element.parentElement;
  }
  return !!element;
};

export const getFirstFocusableDescendant = (element: Element) => {
  const children = element.getElementsByTagName("*");
  for (let i = 0; i < children.length; i += 1) {
    const child = children[i];
    const tagName = child.tagName.toLowerCase();
    if (tagName === "input" || tagName === "select") {
      return child;
    }
  }
  return undefined;
};

export const getBodyChildElements = () => {
  const excludeMatch = /^(script|link)$/i;
  const children: HTMLElement[] = [];
  [].forEach.call(document.body.children, (node: HTMLElement) => {
    if (!excludeMatch.test(node.tagName)) {
      children.push(node);
    }
  });
  return children;
};

export const getNewContainer = (
  target = document.body,
  targetChildPosition?: "first" | "last"
) => {
  // setup DOM
  const container = document.createElement("div");
  if (targetChildPosition === "first") {
    // for SkipLinks
    target.prepend(container);
  } else {
    target.appendChild(container);
  }
  return container;
};

export const setFocusWithoutScroll = (element: HTMLElement) => {
  const x = window.scrollX;
  const y = window.scrollY;
  element.focus();
  window.scrollTo(x, y);
};

const TABINDEX = "tabindex";
const TABINDEX_STATE = "data-g-tabindex";

export const makeNodeFocusable = (node: HTMLElement) => {
  // do not touch aria live containers so that announcements work
  if (!node.hasAttribute("aria-live")) {
    node.setAttribute("aria-hidden", "false");
    // allow children to receive focus again
    const elements = node.getElementsByTagName("*");
    // only reset elements we've changed in makeNodeUnfocusable()
    Array.prototype.filter
      .call(elements || [], (element) => element.hasAttribute(TABINDEX_STATE))
      .forEach((element) => {
        const prior = element.getAttribute(TABINDEX_STATE);
        if (prior >= 0) {
          element.setAttribute(TABINDEX, element.getAttribute(TABINDEX_STATE));
        } else if (prior === "none") {
          element.removeAttribute(TABINDEX);
        }
        element.removeAttribute(TABINDEX_STATE);
      });
  }
};

const autoFocusingTags = /(a|area|input|select|textarea|button|iframe)$/;

export const makeNodeUnfocusable = (node: HTMLElement) => {
  // do not touch aria live containers so that announcements work
  if (!node.hasAttribute("aria-live")) {
    node.setAttribute("aria-hidden", "true");
    // prevent children to receive focus
    const elements = node.getElementsByTagName("*");
    // first, save off the tabIndex of any element with one
    Array.prototype.filter
      .call(
        elements || [],
        (element) => element.getAttribute(TABINDEX) !== null
      )
      .forEach((element) => {
        element.setAttribute(TABINDEX_STATE, element.getAttribute(TABINDEX));
        element.setAttribute(TABINDEX, -1);
      });
    // then, if any element is inherently focusable and not handled above,
    // give it a tabIndex of -1 so it can't receive focus
    Array.prototype.filter
      .call(elements || [], (element) => {
        const currentTag = element.tagName.toLowerCase();
        return (
          currentTag.match(autoFocusingTags) &&
          element.focus &&
          element.getAttribute(TABINDEX_STATE) === null
        );
      })
      .forEach((element) => {
        element.setAttribute(TABINDEX_STATE, "none");
        element.setAttribute(TABINDEX, -1);
      });
  }
};

export const findVisibleParent = (element: HTMLElement): any => {
  if (element) {
    return element.offsetParent
      ? element
      : findVisibleParent(element.parentElement as HTMLElement) || element;
  }
  return undefined;
};

export const isNodeAfterScroll = (node: HTMLElement, target: HTMLElement) => {
  const { bottom } = node.getBoundingClientRect();
  // target will be the document from findScrollParent()
  const { height, top } = target.getBoundingClientRect
    ? target.getBoundingClientRect()
    : { height: 0, top: 0 };
  return bottom >= top + height;
};

export const isNodeBeforeScroll = (node: HTMLElement, target: HTMLElement) => {
  const { top } = node.getBoundingClientRect();
  // target will be the document from findScrollParent()
  const { top: targetTop } = target.getBoundingClientRect
    ? target.getBoundingClientRect()
    : { top: 0 };
  return top <= targetTop;
};
