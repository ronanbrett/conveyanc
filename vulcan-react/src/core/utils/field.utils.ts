import { FieldMetaProps } from "formik";

export const addFieldValidationClasses = (
  formikMeta: Partial<FieldMetaProps<any>>
) => {
  return `${
    formikMeta.error && formikMeta.touched ? "field--invalid" : "field--valid"
  }`;
};
