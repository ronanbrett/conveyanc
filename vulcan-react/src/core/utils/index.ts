export * from "./dom.utils";
export * from "./field.utils";
export * from "./mapChildren.util";
export * from "./portal.utils";
export * from "./refs.utils";
export * from "./rxjs.utils";
export * from "./formik-effect.util";
export * from "./debounce.util";
export * from "./throttle.util";
