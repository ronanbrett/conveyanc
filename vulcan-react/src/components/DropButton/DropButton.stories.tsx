import React from "react";
import { IconButton } from "../index";
import DropButton from "./DropButton";

const DropContent = ({ onClose }) => (
  <div style={{ padding: "16px" }}>
    <div
      style={{
        display: "flex",
        flexDirection: "row",
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      <h3>Heading</h3>

      <IconButton onClick={onClose} icon="close"></IconButton>
    </div>
    <p>Content</p>
  </div>
);

const SimpleDropButton = () => {
  const [open, setOpen] = React.useState<boolean>();
  const onOpen = () => {
    setOpen(true);
  };
  const onClose = () => {
    setOpen(false);
  };

  return (
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        width: "100vw",
        height: "100vh",
      }}
    >
      <div
        style={{
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
        }}
      >
        <DropButton
          open={open}
          onOpen={onOpen}
          onClose={onClose}
          dropContent={<DropContent onClose={onClose} />}
          dropProps={{ align: { top: "bottom", right: "left" } }}
        >
          Open
        </DropButton>
      </div>
    </div>
  );
};

export default {
  title: "Components/DropButton",
  decorators: [],
  component: DropButton,
};

export const Default = (props) => <SimpleDropButton />;

export const IsOPen = (props) => <SimpleDropButton open={true} />;
