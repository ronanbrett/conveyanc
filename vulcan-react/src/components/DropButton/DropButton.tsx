import { Button, Drop } from "@components";
import { useForwardedRef } from "@core/utils/refs.utils";
import React, {
  forwardRef,
  MutableRefObject,
  useCallback,
  useEffect,
  useState,
} from "react";
import { DropProps } from "../Drop/DropContainer";
import classNames from "classnames";
import "./DropButton.scss";

export interface DropButtonProps {
  a11yTitle: string;
  dropAlign?: {
    top?: "top" | "bottom";
    bottom?: "top" | "bottom";
    right?: "left" | "right";
    left?: "left" | "right";
  };
  disabled: boolean;
  dropContent: JSX.Element;
  dropTarget?: HTMLElement;
  dropProps?: DropProps;
  id: string;
  onClose?: (...args: any[]) => void;
  onOpen?: (...args: any[]) => void;
  onClick?: (...args: any[]) => void;
  open?: boolean;
  [param: string]: any;
}

const DropButton = forwardRef(
  (
    {
      a11yTitle = "Open Drop",
      disabled,
      dropAlign = { top: "bottom", left: "left" },
      dropProps,
      dropContent,
      dropTarget,
      id,
      open,
      onClick,
      onClose,
      onOpen,
      ...rest
    }: DropButtonProps,
    ref
  ) => {
    console.log(dropAlign);

    const buttonRef = useForwardedRef(ref) as MutableRefObject<any>;
    const [show, setShow] = useState<boolean>();
    useEffect(() => {
      if (open !== undefined && open !== show) {
        setShow(open);
      }
    }, [open, show]);

    const onDropClose = useCallback(
      (event) => {
        // if the user has clicked on our Button, don't do anything here,
        // handle that in onClickInternal() below.
        let node = event.target;
        while (node !== document && node !== buttonRef.current) {
          node = node.parentNode;
        }
        if (node !== buttonRef.current) {
          // don't change internal state if caller is driving
          if (open === undefined) setShow(false);
          if (onClose) onClose(event);
        }
      },
      [buttonRef, onClose, open]
    );

    const onClickInternal = useCallback(
      (event) => {
        if (!show) {
          setShow(true);
          if (onOpen) onOpen(event);
        } else {
          setShow(false);
          if (onClose) onClose(event);
        }
        if (onClick) onClick(event);
      },
      [onClick, onClose, onOpen, show]
    );

    const cx = classNames({
      DropButton: true,
      opened: open,
    });

    return (
      <>
        <Button
          id={id}
          classes={cx}
          ref={buttonRef}
          a11yTitle={a11yTitle}
          disabled={disabled}
          {...rest}
          onClick={onClickInternal}
        />
        {show && buttonRef.current && (
          <Drop
            id={id ? `${id}__drop` : undefined}
            restrictFocus
            align={dropAlign}
            target={dropTarget || buttonRef.current}
            onClickOutside={onDropClose}
            onEsc={onDropClose}
            {...dropProps}
          >
            {dropContent}
          </Drop>
        )}
      </>
    );
  }
);
export default DropButton;
