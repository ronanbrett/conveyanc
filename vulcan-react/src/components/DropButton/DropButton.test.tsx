import React from "react";
import { cleanup, fireEvent, render } from "@testing-library/react";
import DropButton from "./DropButton";
import { create } from "react-test-renderer";


import { axe } from "jest-axe";
import "jest-axe/extend-expect";

import { createPortal, expectPortal } from "@core/utils/portal.utils";
describe("DropButton", () => {
  beforeEach(createPortal);

  afterEach(cleanup);

  test("should have no accessibility violations", async () => {
    const { container } = render(
      <DropButton
        a11yTitle="test"
        dropContent={<div id="drop-contents">drop contents</div>}
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("closed", () => {
    const component = create(
      <DropButton
        label="Dropper"
        dropContent={<div id="drop-contents">drop contents</div>}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("opened", () => {
    const component = create(
      <DropButton
        id="123"
        open={true}
        dropContent={<div id="drop-contents">drop contents</div>}>
           Dropper
           </DropButton>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('open and close', () => {
   window.scrollTo = jest.fn();
   const { getByText, container } = render(
     <DropButton
       dropContent={<div id="drop-contents">Drop Contents</div>}>
          Dropper
       </DropButton>,
   );
   expect(container.firstChild).toMatchSnapshot();
   expect(document.getElementById('drop-contents')).toBeNull();

   fireEvent.click(getByText('Dropper'));
   expectPortal('drop-contents').toMatchSnapshot();

   fireEvent.click(getByText('Dropper'));
   expect(document.getElementById('drop-contents')).toBeNull();
   expect(window.scrollTo).toBeCalled();
 });

 test('close by clicking outside', done => {
   const { getByText, container } = render(
     <DropButton
       dropContent={<div id="drop-contents">Drop Contents</div>}
     >Dropper</DropButton>,
   );
   expect(container.firstChild).toMatchSnapshot();
   expect(document.getElementById('drop-contents')).toBeNull();

   fireEvent.click(getByText('Dropper'));
   expectPortal('drop-contents').toMatchSnapshot();

   fireEvent(
     document,
     new MouseEvent('mousedown', { bubbles: true, cancelable: true }),
   );

   setTimeout(() => {
     expect(document.getElementById('drop-contents')).toBeNull();
     done();
   }, 50);
 });

 test('disabled', () => {
   const { getByText, container } = render(
     <DropButton
       disabled
       dropContent={<div id="drop-contents">Drop Contents</div>}
     >Dropper</DropButton>,
   );
   expect(container.firstChild).toMatchSnapshot();
   expect(document.getElementById('drop-contents')).toBeNull();

   fireEvent.click(getByText('Dropper'));
   expect(document.getElementById('drop-contents')).toBeNull();
 });

 test('opened ref', () => {
   const ref = React.createRef();
   const { container } = render(
     <DropButton
       ref={ref}
       open
       dropContent={<div id="drop-contents">Drop Contents</div>}
     >Dropper</DropButton>,
   );
   expect(container.firstChild).toMatchSnapshot();
   expectPortal('drop-contents').toMatchSnapshot();
 });

 test('ref function', () => {
   const ref = jest.fn();
   const { container } = render(
     <DropButton
       ref={ref}
       open
       dropContent={<div id="drop-contents">Drop Contents</div>}
     >Dropper</DropButton>,
   );
   expect(container.firstChild).toMatchSnapshot();
   expect(ref).toBeCalled();
   expectPortal('drop-contents').toMatchSnapshot();
 });

});
