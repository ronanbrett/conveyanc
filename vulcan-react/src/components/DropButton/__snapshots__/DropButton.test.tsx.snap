// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropButton close by clicking outside 1`] = `
<button
  aria-label="Open Drop"
  class="Button Button--primary Button--medium focus"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton close by clicking outside 2`] = `
<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton closed 1`] = `
<button
  aria-label="Open Drop"
  className="Button Button--primary Button--medium focus"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  type="button"
/>
`;

exports[`DropButton disabled 1`] = `
<button
  aria-label="Open Drop"
  class="Button Button--primary Button--medium focus"
  disabled=""
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton open and close 1`] = `
<button
  aria-label="Open Drop"
  class="Button Button--primary Button--medium focus"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton open and close 2`] = `
<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton opened 1`] = `
<button
  aria-label="Open Drop"
  className="Button Button--primary Button--medium focus"
  id="123"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton opened ref 1`] = `
<button
  aria-label="Open Drop"
  class="Button Button--primary Button--medium focus"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton opened ref 2`] = `
<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton ref function 1`] = `
<button
  aria-label="Open Drop"
  class="Button Button--primary Button--medium focus"
  type="button"
>
  Dropper
</button>
`;

exports[`DropButton ref function 2`] = `
<div
  id="drop-contents"
>
  Drop Contents
</div>
`;

exports[`DropButton should have no accessibility violations 1`] = `
<button
  aria-label="test"
  class="Button Button--primary Button--medium focus"
  type="button"
/>
`;
